### PR TITLE
npm build: create addons directory that loads from lib. Closes #3780

### LINKED ISSUE #3780: Expose React.addons.* as react/addons/* on npm and deprecate react/addons
The current system of packaging react/addons isn't great because you need to pull in all the addons even if you just want to use one of them. Instead, let's make it so you can do `require('react/addons/TransitionGroup')` and similar.

This should be as simple as adding a few files to build/npm-react that require the (private) internal modules.

```js
// npm-react/addons/TransitionGroup.js
module.exports = require('../lib/ReactTransitionGroup.js');
```

### PR DESCRIPTION
**Pull request by iamdustan**
This generates the following files:

screen shot 2015-04-29 at 4 10 22 pm

### CODE REVIEW

**Comment by zpao** [comment]
I actually think we should match up with the fields on `React.addons` now. I don't think we'll be able to fully automate this. eg, `react/addons/createFragment` needs to look like `module.exports = require('../lib/ReactFragment').create;`

**Comment by iamdustan** [comment]
Ah. So looking at `build/npm-react/lib/ReactWithAddons.js`, it looks like the following are the not-straight export ones:

```
CSSTransitionGroup: ReactCSSTransitionGroup,
PureRenderMixin: ReactComponentWithPureRenderMixin,
TransitionGroup: ReactTransitionGroup,
batchedUpdates: ReactUpdates.batchedUpdates
createFragment: ReactFragment.create,
Perf: ReactDefaultPerf,
TestUtils: ReactTestUtils,
```

Dang. Three items in and I realized it would’ve been easier to have the default ones done. What is easier for your team for maintenance going forward? Just manually writing in the npm build source the mapping or look into using `esprima-fb` for the transform.

Should be pretty straightforward.

Visit all `React.addons` keys. Create a new file with the key name with the value.

**Comment by zpao** [comment]
I liked @spicyj's idea of just adding the files directly into the repo. Though I guess he said "build/npm-react" which isn't right and I think he miswrote. We can do it before the build process starts.

If we create an addons folder in https://github.com/facebook/react/tree/master/npm-react with the right files and the right content, we don't have to do anything fancy. It's more prone to falling out of date but perhaps we can do some simple testing.

**Comment by spicyj** [comment]
Yeah, my bad – I meant just npm-react/.

**Comment by iamdustan** [comment]
...and updated. Is this what you were looking for?

Also, I added a deprecation notice for `require("react/addons");` in a similar style to other `deprecated` notifications I found in the code. I think that should only be called once in userland since require calls are cached, so I didn’t wrap it.

**Comment by spicyj** [comment]
I think these all need to be `../lib` not `./lib`. Were you able to test these at all? If you run grunt npm-react:pack then you should be able to install from build/react.tgz.

**Comment by zpao** [comment]
That's it (except for the paths as Ben mentioned) :)

A couple things I'm not sure about:
1. The deprecation notice. It might end up being that we really do want to deprecate this but I'm not entirely sure yet.
2. What to do about `__DEV__`. Right now `React.addons.TestUtils` doesn't exist at all in prod. Same with `Perf`. Now maybe it doesn't matter so much since people would be requiring explicitly but it is a slight difference.

**Comment by spicyj** [comment]
I'm personally up for deprecating react/addons now (though I'd probably do a console.warn) and I don't think the `__DEV__` difference is that important here.

**Comment by iamdustan** [comment]
Ah. When I first searched for how other deprecations were being used I ended up in the JSXTransformer. In the React source I’m seeing the `warning` module used. Should I use that instead?

https://github.com/facebook/react/search?l=javascript&q=%22is+deprecated%22&type=Code&utf8=%E2%9C%93

I have successfully installed from the .tgz file. Doing a quick, manual sanity check right now.

**Comment by iamdustan** [comment]
Looks legit to me:

screen shot 2015-04-29 at 5 30 53 pm

**Comment by iamdustan** [comment]
>  2. What to do about __DEV__. Right now React.addons.TestUtils doesn't exist at all in prod. Same with Perf. Now maybe it doesn't matter so much since people would be requiring explicitly but it is a slight difference.

I really like explicit change. It is very attractive to be able to move my team to `var TestUtils = require('react/addons/TestUtils)` and disallow requiring the entire `react/addons` in app code.

File: npm-react/addons.js [hunk 1]
```diff
@@ -1 +1,8 @@
+var warning = require('./lib/warning');
+warning(
+  false,
+  'require("react/addons") is deprecated. ' +
+  'Access using require("react/addons/{addon}" instead.'
```
Comment (spicyj) [inline npm-react/addons.js]: Missing paren here. Otherwise looks good to me – mind squashing your commits?

**Comment by iamdustan** [comment]
Done.

File: npm-react/addons.js [hunk 2]
```diff
@@ -1 +1,8 @@
+var warning = require('./lib/warning');
+warning(
+  false,
+  'require("react/addons") is deprecated. ' +
+  'Access using require("react/addons/{addon})" instead.'
```
Comment (spicyj) [inline npm-react/addons.js]: `")` not `)"` :)

**Comment by iamdustan** [comment]
Doh! I swear I’m paying attention. :innocent:

**Comment by spicyj** [comment]
Thank you!

**Comment by zpao** [comment]
Thanks Dustan!

**Comment by iamdustan** [comment]
Np! I’ll take these 30 minute `[Good First Bug]`s any day!

**Pull request merged**